pkg/segment/segexecution.go

```diff
@@ -157,52 +157,99 @@ func HelperQueryArithmeticAndLogical(queryOps []structs.QueryArithmetic, resMap
+					rGroupID = resultRHS.MetricName + labelStr
+				}
+
+				if queryOp.Operation == utils.LetOr || queryOp.Operation == utils.LetUnless {
+					labelStrSet[labelStr] = struct{}{}
+				}
+
+				if _, ok := resultRHS.Results[rGroupID]; !ok && queryOp.Operation != utils.LetOr && queryOp.Operation != utils.LetUnless {
 					continue
 				} //Entries for which no matching entry in the right-hand vector are dropped
-				finalResult[groupID] = make(map[uint32]float64)
+				finalResult[lGroupID] = make(map[uint32]float64)
 				for timestamp, valueLHS := range tsLHS {
-					valueRHS := resultRHS.Results[groupID][timestamp]
+					valueRHS := resultRHS.Results[rGroupID][timestamp]
 					switch queryOp.Operation {
 					case utils.LetAdd:
-						finalResult[groupID][timestamp] = valueLHS + valueRHS
+						finalResult[lGroupID][timestamp] = valueLHS + valueRHS
 					case utils.LetDivide:
 						if valueRHS == 0 {
 							continue
 						}
-						finalResult[groupID][timestamp] = valueLHS / valueRHS
+						finalResult[lGroupID][timestamp] = valueLHS / valueRHS
 					case utils.LetMultiply:
-						finalResult[groupID][timestamp] = valueLHS * valueRHS
+						finalResult[lGroupID][timestamp] = valueLHS * valueRHS
 					case utils.LetSubtract:
-						finalResult[groupID][timestamp] = valueLHS - valueRHS
+						finalResult[lGroupID][timestamp] = valueLHS - valueRHS
 					case utils.LetGreaterThan:
 						if valueLHS > valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetGreaterThanOrEqualTo:
 						if valueLHS >= valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetLessThan:
 						if valueLHS < valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetLessThanOrEqualTo:
 						if valueLHS <= valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetEquals:
 						if valueLHS == valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetNotEquals:
 						if valueLHS != valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
+					case utils.LetAnd:
+						finalResult[lGroupID][timestamp] = valueLHS
+					case utils.LetOr:
+						finalResult[lGroupID][timestamp] = valueLHS
+					case utils.LetUnless:
+						finalResult[lGroupID][timestamp] = valueLHS
+					}
+				}
+			}
+			if queryOp.Operation == utils.LetOr || queryOp.Operation == utils.LetUnless {
+				for rGroupID, tsRHS := range resultRHS.Results {
+					labelStr := ""
+					if len(rGroupID) >= len(resultLHS.MetricName) {
+						labelStr = rGroupID[len(resultLHS.MetricName):]
+					}
```

Comment:
This seems wrong? Why are you doing `rGroupID[len(resultLHS.MetricName):]`? Won't that give bad results if the LHS and RHS have different-length metric names?

Reply:
Yes, that's a bug. I just want to extract label set by remove the metricName at the beginning